### PR TITLE
Make positionals opt-in when strict:true

### PR DESCRIPTION
**Pull request by bakkot**
Fixes https://github.com/pkgjs/parseargs/issues/85.

In the second commit I've updated the message for unknown options to suggest using `--`. I'm not sure if this is worth doing, but I figured I'd implement it for discussion; happy to revert if it seems unhelpful.

### CODE REVIEW

**Comment by shadowspawn** [comment]
From https://github.com/pkgjs/parseargs/issues/85#issuecomment-1094143617

> That is, I think that making positionals opt-in is in keeping with the "you will need to provide config for each option that you want to accept" design. I regard positionals as being a kind of option that you want to accept.

I think giving authors insight into the design helps smooth the experience, so more likely the API works the way they expect it to work because we guided their expectations (https://github.com/pkgjs/parseargs/issues/85#issuecomment-1098832809).

I am interested in seeing a sentence or two in the docs to spin the config-what-you-accept story or similar. I was thinking outside the documentation of the properties themselves and tying in the options.

**Review by ljharb — APPROVED** [review]

**Comment by bakkot** [comment]
> I am interested in seeing a sentence or two in the docs to spin the config-what-you-accept story or similar. I was thinking outside the documentation of the properties themselves and tying in the options.

Makes sense. I've added the following sentence as a start, though I'm happy to try to expand this section:

> Takes a specification for the expected options and returns a structured object corresponding to passed arguments.

File: index.js [hunk 1]
```diff
@@ -141,11 +142,13 @@ function storeOption(longOption, optionValue, options, values) {
 const parseArgs = (config = { __proto__: null }) => {
   const args = objectGetOwn(config, 'args') ?? getMainArgs();
   const strict = objectGetOwn(config, 'strict') ?? true;
+  const allowPositionals = objectGetOwn(config, 'allowPositionals') ?? !strict;
```
Comment (shadowspawn) [inline index.js:145]: A brief comment on change in previous design/expectations.

This means `parseArgs` may now throw in `strict:false` mode for userArgs for first time. But it is opt-in and I think make sense when write it first time, and makes sense to read later.

Comment (bakkot) [reply index.js:145]: Yeah, I thought this was better than making it so you could have an explicit `allowPositionals: false` but still get positionals. The alternative is to reject `allowPositionals: false` in combination with `strict: false`, but I didn't see much value in that.

File: README.md [hunk 2]
```diff
@@ -38,7 +40,7 @@ added: REPLACEME
   * `positionals` {string\[]}, containing positional arguments.
 
 Provides a higher level API for command-line argument parsing than interacting
-with `process.argv` directly.
+with `process.argv` directly. Takes a specification for the expected options and returns a structured object corresponding to passed arguments.
```
Comment (shadowspawn) [inline README.md:43]: Trying a variation to cover options and positionals. I wanted to try and spin the story the config is for everything you expect.

This does give positional more weight than has in the config, but values/positionals are more equally righted in  the result.

```suggestion
with `process.argv` directly. Takes a specification for the expected arguments and returns a structured object with the parsed options and positionals.
```

Comment (bakkot) [reply README.md:43]: SGTM

**Comment by shadowspawn** [comment]
Optional: there could be a single check for `ERR_PARSE_ARGS_UNEXPECTED_POSITIONAL` after parsing, rather than two checks inside the loop.

**Review by shadowspawn — APPROVED** [review]
One optional suggestion. Nice tests.

**Comment by bakkot** [comment]
> Optional: there could be a single check for `ERR_PARSE_ARGS_UNEXPECTED_POSITIONAL` after parsing, rather than two checks inside the loop.

Doing the check after parsing means any other errors will happen first, even errors which are later in the input. I prefer getting errors in the order in which the corresponding problematic things appear.

File: index.js [hunk 3]
```diff
@@ -86,12 +87,12 @@ function getMainArgs() {
  * @param {boolean} strict - show errors, from parseArgs({ strict })
```
Comment (shadowspawn) [inline index.js:87]: Nit: missing JSDoc for new parameter.

**Review by bcoe — APPROVED** [review]
LGTM, thank you for the contribution 👍

**Comment by shadowspawn** [comment]
> In the second commit I've updated the message for unknown options to suggest using --.  I'm not sure if this is worth doing, but I figured I'd implement it for discussion; happy to revert if it seems unhelpful.

Something for future, not urgent.

We didn't really have time to discuss this. I expect most errors for unknown options will be for typos in the option name, so the suggestion is not going to be relevant. Not harmful, but not helpful. We are going the extra mile with error messages since they may be the UX, so could perhaps have two suggestions, with a "check the spelling" type suggestion first?

**Pull request merged**